cluster/hosts.go

```diff
@@ -54,14 +56,37 @@ func (c *Cluster) TunnelHosts(ctx context.Context, flags ExternalFlags) error {
 		return err
 	}
 	for _, host := range c.InactiveHosts {
-		log.Warnf(ctx, "Removing host [%s] from node lists", host.Address)
-		c.EtcdHosts = removeFromHosts(host, c.EtcdHosts)
-		c.ControlPlaneHosts = removeFromHosts(host, c.ControlPlaneHosts)
-		c.WorkerHosts = removeFromHosts(host, c.WorkerHosts)
-		c.RancherKubernetesEngineConfig.Nodes = removeFromRKENodes(host.RKEConfigNode, c.RancherKubernetesEngineConfig.Nodes)
+		c.removeHostFromNodeList(ctx, host)
 	}
 	return ValidateHostCount(c)
+}
+
+func (c *Cluster) RemoveHostsLabeledToIgnoreUpgrade(ctx context.Context) {
+	kubeClient, err := k8s.NewClient(c.LocalKubeConfigPath, c.K8sWrapTransport)
+	nodes, err := kubeClient.CoreV1().Nodes().List(metav1.ListOptions{})
+	if err != nil {
+		// not returning error since upgrade can proceed
+		log.Infof(ctx, "[RemoveHostsLabeledToIgnoreUpgrade] Error listing nodes: %v", err)
+	}
+	if nodes == nil {
+		return
+	}
+	for _, node := range nodes.Items {
+		if val, ok := node.Labels[k8s.IgnoreHostDuringUpgradeLabel]; ok && val == "true" {
+			host := hosts.Host{RKEConfigNode: v3.RKEConfigNode{Address: node.Annotations[k8s.ExternalAddressAnnotation]}}
+			c.removeHostFromNodeList(ctx, &host)
+			c.InactiveHosts = removeFromHosts(&host, c.InactiveHosts)
+		}
+	}
+	return
+}
 
+func (c *Cluster) removeHostFromNodeList(ctx context.Context, host *hosts.Host) {
+	log.Warnf(ctx, "Removing host [%s] from node lists", host.Address)
+	c.EtcdHosts = removeFromHosts(host, c.EtcdHosts)
```

Comment:
Can't this be consolidated into a function removeFromAllHosts and pass it once?

Reply:
sure,  this function existed prior to my changes but will try

Reply:
This function existed prior to my change, also looks like it's being called for each role separately since each of the lists etcdHosts, ControlHosts and workerHosts should get updated